### PR TITLE
Simplify handling of upgrade information.

### PR DESCRIPTION
**Pull request by twittner**
The current `UpgradeInfo` trait defines the way to denote upgrade meta information, most importantly the supported protocol names:

```rust
pub trait UpgradeInfo {
    type UpgradeId;
    type NamesIter: Iterator<Item = (Bytes, Self::UpgradeId)>;

    fn protocol_names(&self) -> Self::NamesIter;
}
```

As issue #660 noted, most protocols use static byte slices and incur unnecessary allocations during each iteration by being forced to return `Bytes`. This PR moves the meta-information into the respective upgrade traits and defines the interface as follows:

```rust
pub trait {Inbound,Outbound}Upgrade {
    ...
    type Info: AsRef<[u8]> + Clone;
    type InfoIter: IntoIterator<Item = Self::Info>;
    
    fn info_iter(&self) -> Self::InfoIter;
    ...
}
```

So, instead of names and upgrade IDs we use a single associated type `Info` which must support a byte-slice view. In `upgrade_{inbound,outbound}` the actual type is passed in allowing the upgrade to embed arbitrary information but for most upgrades, `Info` will just be `&'static [u8]`, e.g. for `mplex`:

```rust
impl<C> InboundUpgrade<C> for MplexConfig
where
    C: AsyncRead + AsyncWrite,
{
    ...
    type Info = &'static [u8];
    type InfoIter = iter::Once<Self::Info>;

    fn info_iter(&self) -> Self::InfoIter {
        iter::once(b"/mplex/6.7.0")
    }
    ...
```

There is no heap allocation involved and while this PR does not fully exhaust the optimisation possibilities a subsequent one will (cf. [here](https://github.com/twittner/rust-libp2p/compare/upgrade-names...twittner:optimise-multistream-select) for the work-in-progress).

As a further simplification, this PR removes the explicit passing of a custom equality predicate as part of the iteration during upgrade application in preference of regular equality comparison.

### CODE REVIEW

File: core/src/either.rs [hunk 1]
```diff
@@ -342,3 +342,14 @@ where
     }
 }
 
+#[derive(Debug, Clone)]
+pub enum EitherInfo<A, B> { A(A), B(B) }
```
Comment (tomaka) [inline core/src/either.rs]: Another either :no_mouth:

**Comment by tomaka** [comment]
I have two remarks:

- I quite liked that there was an `UpgradeInfo` trait, as you didn't have to specify the type of the substream in order to query the protocol names of an upgrade. The fact that it was the same trait for both inbound and outbound also reduced code duplication and forced implementations to support the same protocols on both sides. I'm a bit confused as to why you removed the trait.
- Instead of `AsRef<[u8]> + Clone`, maybe a `ProtocolName` trait or something? I'm not sure.

**Comment by twittner** [comment]
> @tomaka: The fact that it was the same trait for both inbound and outbound also reduced code duplication and forced implementations to support the same protocols on both sides. I'm a bit confused as to why you removed the trait.

I agree w.r.t. the code duplication. The change was motivated because I wanted to support different upgrades for inbound and outbound direction. After all, protocols can decide to be unidirectional by implementing only one of `InboundUpgrade` and `OutboundUpgrade` and I wanted a way to be able to compose protocols together such that the end-result could be passed to `Transport::with_upgrade` (which requires `InboundUpgrade` + `OutboundUpgrade`). In order to do that I had a builder in mind, roughly along the lines of:

```rust
pub struct UpgradeBuilder<S, I, O> {
    inbound: I,
    outbound: O,
    _witness: PhantomData<S>
}

impl<S, I, O> UpgradeBuilder<S, I, O> where I: InboundUpgrade<S> {
    pub fn or_inbound<T>(self, t: T) -> UpgradeBuilder<S, SelectUpgrade<I, T>, O>
    where
        T: InboundUpgrade<S>
    {
        UpgradeBuilder {
            inbound: SelectUpgrade::new(self.inbound, t),
            outbound: self.outbound,
            _witness: PhantomData
        }
    }
}

// similar for `UpgradeBuilder::or_outbound`
```

This builder would then implement `InboundUpgrade` and `OutboundUpgrade` but with only a single `UpgradeInfo` it could not dispatch its own `UpgradeInfo` impl to the respective `self.inbound` or `self.outbound` branch. But that was just an idea. If the `UpgradeBuilder` is not desired I can easily bring back the `UpgradeInfo` trait.

> @tomaka: Instead of `AsRef<[u8]> + Clone`, maybe a `ProtocolName` trait or something?

I thought the same thing. However the trait `ProtocolName` would need to be used by `multistream-select` during protocol negotiation, which would require a circular dependency from `multistream-select` to `libp2p_core` if we put it into `libp2p_core`. If we are happy to put such a trait into `multistream-select` and expose it here as a constraint on `Info`, I could make the change.

> @dvdplm: I think "info" is an inflated word and things like `info_iter().into_iter()` is a bit hard on the eyes. Maybe `protocol_info` or even `protocol_meta_info`?

I am fine with changing names as long as they do not get too long.

File: core/src/upgrade/apply.rs [hunk 2]
```diff
@@ -204,38 +204,30 @@ where
     }
 }
 
-/// Wraps around a `UpgradeInfo` and satisfies the requirement of `listener_select_proto`.
-struct UpgradeIntoProtocolsIterWrap<U>(U);
+// Wraps around a `UpgradeInfo` and satisfies the requirement of `listener_select_proto`.
```
Comment (tomaka) [inline core/src/upgrade/apply.rs]: Nit; something being private doesn't mean we should doc-comment it; the doc-comment won't show up in the docs but can still help people developers using the RLS and are hacking on libp2p
```suggestion
/// Wraps around a `UpgradeInfo` and satisfies the requirement of `listener_select_proto`.
```

File: core/src/upgrade/mod.rs [hunk 3]
```diff
@@ -76,18 +75,28 @@ pub use self::{
...
 };
 
+/// Types serving as protocol names.
+pub trait ProtocolName {
+    /// The protocol name as bytes.
+    fn protocol_name(&self) -> &[u8];
+}
+
+impl<T: AsRef<[u8]>> ProtocolName for T {
+    fn protocol_name(&self) -> &[u8] {
```
Comment (tomaka) [inline core/src/upgrade/mod.rs]:
```suggestion
    #[inline]
    fn protocol_name(&self) -> &[u8] {
```

File: misc/multistream-select/src/lib.rs [hunk 4]
```diff
@@ -55,17 +55,12 @@
 //! let client = TcpStream::connect(&"127.0.0.1:10333".parse().unwrap())
 //!     .from_err()
 //!     .and_then(move |connec| {
-//!         let protos = vec![
-//!             (Bytes::from("/echo/1.0.0"), <Bytes as PartialEq>::eq, MyProto::Echo),
-//!             (Bytes::from("/hello/2.5.0"), <Bytes as PartialEq>::eq, MyProto::Hello),
-//!         ]
-//!                         .into_iter();
+//!         let protos = vec![b"/echo/1.0.0", b"/echo/2.5.0"];
```
Comment (tomaka) [inline misc/multistream-select/src/lib.rs]: :+1:

**Pull request merged**